AFQ/tasks/tractography.py

```diff
@@ -134,14 +142,109 @@ def streamlines(data_imap, seed, stop,
+        num_chunks = multiprocessing.cpu_count() - 1
+
     if is_trx:
         start_time = time()
         dtype_dict = {'positions': np.float16, 'offsets': np.uint32}
-        lazyt = aft.track(params_file, **this_tracking_params)
-        sft = TrxFile.from_lazy_tractogram(
-            lazyt,
-            seed,
-            dtype_dict=dtype_dict)
+        if num_chunks and num_chunks > 1:
+            if not has_ray:
+                raise ImportError("Ray is required to perform tractography in"
+                                  "parallel, please install ray or remove the"
+                                  " 'num_chunks' arg")
+
+            @ray.remote
+            class TractActor():
+                def __init__(self):
+                    self.TrxFile = TrxFile
+                    self.aft = aft
+                    self.objects = {}
+
+                def trx_from_lazy_tractogram(self, lazyt_id, seed, dtype_dict):
+                    id = self.objects[lazyt_id]
+                    return self.TrxFile.from_lazy_tractogram(
+                        id,
+                        seed,
+                        dtype_dict=dtype_dict)
+
+                def create_lazyt(self, id, *args, **kwargs):
+                    self.objects[id] = self.aft.track(*args, **kwargs)
+                    return id
+
+                def delete_lazyt(self, id):
+                    if id in self.objects:
+                        del self.objects[id]
+            actors = [TractActor.remote() for _ in range(num_chunks)]
+            object_id = 1
+            tracking_params_list = []
+
+            # random seeds case
+            if isinstance(this_tracking_params.get("n_seeds"), int) and \
+               this_tracking_params.get("random_seeds"):
+
+                remainder = this_tracking_params['n_seeds'] % num_chunks
+                for i in range(num_chunks):
+                    # create copy of tracking params
+                    copy = this_tracking_params.copy()
+                    n_seeds = this_tracking_params['n_seeds']
+                    copy['n_seeds'] = n_seeds // num_chunks
+                    # add remainder to 1st list
+                    if i == 0:
+                        copy['n_seeds'] += remainder
+                    tracking_params_list.append(copy)
+
+            elif isinstance(this_tracking_params.get("n_seeds"), (np.ndarray,
+                                                                  list)):
+                n_seeds = np.array(this_tracking_params['n_seeds'])
+                seed_chunks = np.array_split(n_seeds, num_chunks)
+                tracking_params_list = [this_tracking_params.copy() for _ in
+                                        range(num_chunks)]
+
+                for i in range(num_chunks):
+                    tracking_params_list[i]['n_seeds'] = seed_chunks[i]
+
+            else:
+                seeds = gen_seeds(
+                    this_tracking_params.get('seed_mask', None),
```

Comment:
You don't need the get() calls here either, as this_tracking_params should already be filled out previously in the task system. So these can be:
```
this_tracking_params['seed_mask']
```
etc.

Reply:
good to know thanks! Switched them to [] index's in the most recent commit